proto_parser/src/proto_value_test.cc

```diff
@@ -22,8 +22,183 @@
 #include <google/protobuf/util/message_differencer.h>
 
 namespace wireless_android_play_analytics{
+ 
+const std::string text_proto = R"pb(
+# comment 1
+int32_field: 1
 
-// TODO(alexanderlin): Implement.
+# comment 2
+field_nested_message { # comment 3
+  int64_field: 100
+}
+
+field_nested_message {
+  # comment 4
+  # comment 5
+  int64_field: 100
+}
+
+
+bool_field: true
+bool_field: false # comment 6
+)pb";
+
+class ProtoValueTest : public ::testing::Test {
+ protected:
+  
+  virtual void SetUp() {
+    message_ = ProtoValue::Create(text_proto, test_);
+    message_val_ = dynamic_cast<MessageValue*>(message_.get());
+  }
+
+  TestProto test_;
+  std::unique_ptr<ProtoValue> message_;
+  MessageValue* message_val_;
+};
+
+TEST_F(ProtoValueTest, OuterProtoTest) {
+  const std::vector<std::unique_ptr<ProtoValue>>& fields = 
+      message_val_->GetFields();
+  EXPECT_EQ(message_val_->GetName(), "");
+  EXPECT_EQ(message_val_->GetCommentAboveField().size(), 0);
+  EXPECT_EQ(message_val_->GetCommentBehindField(), "");
+  EXPECT_EQ(message_val_->GetIndentCount(), 0);
+  EXPECT_EQ(fields.size(), 5);
+}
+
+TEST_F(ProtoValueTest, FirstFieldTest) {
+  // #comment 1
+  // int32_field: 1
+  const std::vector<std::unique_ptr<ProtoValue>>& fields = 
+      message_val_->GetFields();
+  PrimitiveValue* int32_field = 
+      dynamic_cast<PrimitiveValue*>(fields[0].get());
+  ASSERT_NE(int32_field, nullptr);
+  EXPECT_EQ(int32_field->GetName(), "int32_field");
+  ASSERT_THAT(int32_field->GetCommentAboveField(),
+      testing::ElementsAreArray({"# comment 1",}));
+  EXPECT_EQ(int32_field->GetCommentBehindField(), "");
+
+  // Inititialize the value to make sure it is held by an integer.
+  int int32_field_val = 1;
+  absl::variant<double, float, int, unsigned int, int64_t, uint64_t, bool, 
+      const google::protobuf::EnumValueDescriptor *, std::string> val = 
+      int32_field_val;
+  EXPECT_TRUE(absl::holds_alternative<int>(int32_field->GetVal()));
+  EXPECT_EQ(int32_field->GetVal(), val);
+}
+
+TEST_F(ProtoValueTest, SecondFieldTest) {
+  //  # comment 2
+  // field_nested_message { # comment 3
+  //   int64_field: 100
+  // }
```

Comment:
// Field details (this is a repeated field): 
  //
  //  # comment 2
  // field_nested_message { # comment 3
  //   int64_field: 100
  // }